# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Basics (DataFrame Basics I)

### Exercise 2: Indexing and Slicing

7. __Import Pandas__ (as pd), <br>
__Import__ the Cars Dataset (__cars.csv__) and save the DataFrame in the variable __cars__ and<br>
__Inspect/Print__ cars!

In [1]:
# import pandas
import pandas as pd

In [2]:
# import the dataset
cars = pd.read_csv('cars.csv')

In [4]:
# inspect
print(cars.head())
print(cars.info())
print(cars.describe())

    mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null 

8. Select the column __"name"__ with the __attribute__ (dot) notation! What is the __name__ of the car with __index label 393__?

In [8]:
# cars.head()
cars.name[393]

'ford mustang gl'

The name is ... ford mustang gl!

9. Apply also the __second alternative__ to select the column __"name"__!

In [9]:
cars.loc[393,'name']

'ford mustang gl'

10. __Select__ the car/row at __index position 100__ with the __iloc__ operator (all columns)! How many __cylinders__ does the car have?

In [12]:
cars.iloc[100,:]

mpg                        18
cylinders                   6
displacement              250
horsepower                 88
weight                   3021
acceleration             16.5
model_year                 73
origin                    usa
name            ford maverick
Name: 100, dtype: object

The car has ... 6 cylinders!

11. Select the car/row at __index position 200__ with the __iloc__ operator (and only the column __"name"__)! <br>
What is the __name__ of the car?

In [13]:
cars.iloc[200,:]

mpg                            18
cylinders                       6
displacement                  250
horsepower                     78
weight                       3574
acceleration                   21
model_year                     76
origin                        usa
name            ford granada ghia
Name: 200, dtype: object

It´s a ... ford granada ghia!

12. Select the __last 10 rows__ and the columns __mpg, horsepower, origin__ and __name__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __mpg__ of the __vw pickup__?

In [15]:
cars.iloc[-10:, [0, 3, 7, 8]]

mpg  horsepower  origin                        name
388  26.0        92.0     usa  chrysler lebaron medallion
389  22.0       112.0     usa              ford granada l
390  32.0        96.0   japan            toyota celica gt
391  36.0        84.0     usa           dodge charger 2.2
392  27.0        90.0     usa            chevrolet camaro
393  27.0        86.0     usa             ford mustang gl
394  44.0        52.0  europe                   vw pickup
395  32.0        84.0     usa               dodge rampage
396  28.0        79.0     usa                 ford ranger
397  31.0        82.0     usa              chevrolet s-10

The mpg is ... 44.0!

13. __Import__ the __Cars Dataset__ one more time (save in the variable __cars__) and set the column __"name"__ as the __index__! <br>
__Fill in the Gaps__!

In [16]:
cars = pd.read_csv('cars.csv', index_col= "name")

In [17]:
#run the cell and inspect!
cars

mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0       130.0    3504   
buick skylark 320          15.0          8         350.0       165.0    3693   
plymouth satellite         18.0          8         318.0       150.0    3436   
amc rebel sst              16.0          8         304.0       150.0    3433   
ford torino                17.0          8         302.0       140.0    3449   
...                         ...        ...           ...         ...     ...   
ford mustang gl            27.0          4         140.0        86.0    2790   
vw pickup                  44.0          4          97.0        52.0    2130   
dodge rampage              32.0          4         135.0        84.0    2295   
ford ranger                28.0          4         120.0        79.0    2625   
chevrolet s-10             31.0          4         119.0        82.0    2720   

                           acceleration  model_year  origin  
name                                                         
chevrolet chevelle malibu          12.0          70     usa  
buick skylark 320                  11.5          70     usa  
plymouth satellite                 11.0          70     usa  
amc rebel sst                      12.0          70     usa  
ford torino                        10.5          70     usa  
...                                 ...         ...     ...  
ford mustang gl                    15.6          82     usa  
vw pickup                          24.6          82  europe  
dodge rampage                      11.6          82     usa  
ford ranger                        18.6          82     usa  
chevrolet s-10                     19.4          82     usa  

[398 rows x 8 columns]

In [18]:
#run the cell and inspect!
cars.info()

<class 'pandas.core.frame.DataFrame'>
Index: 398 entries, chevrolet chevelle malibu to chevrolet s-10
Data columns (total 8 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
dtypes: float64(4), int64(3), object(1)
memory usage: 28.0+ KB


14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

In [19]:
cars.index

Index(['chevrolet chevelle malibu', 'buick skylark 320', 'plymouth satellite',
       'amc rebel sst', 'ford torino', 'ford galaxie 500', 'chevrolet impala',
       'plymouth fury iii', 'pontiac catalina', 'amc ambassador dpl',
       ...
       'chrysler lebaron medallion', 'ford granada l', 'toyota celica gt',
       'dodge charger 2.2', 'chevrolet camaro', 'ford mustang gl', 'vw pickup',
       'dodge rampage', 'ford ranger', 'chevrolet s-10'],
      dtype='object', name='name', length=398)

15. Select the __row__ with the index label __"ford ranger"__ (use the __loc__ operator)! What is the __weight__?

In [22]:
cars.loc['ford ranger']

mpg               28
cylinders          4
displacement     120
horsepower        79
weight          2625
acceleration    18.6
model_year        82
origin           usa
Name: ford ranger, dtype: object

The weight is ... 2625!

16. Select the __row__ with the index label __"ford torino"__ and only the columns __"model_year"__ and __"origin"__! <br>
What´s the __model year__?

In [23]:
cars.loc['ford ranger',['model_year', 'origin']]

model_year     82
origin        usa
Name: ford ranger, dtype: object

The model year is ... 20!

17. Select the columns __"mpg"__ and __"weight"__ (all rows) with the __loc__ operator!

In [24]:
cars.loc[:, ["mpg", 'weight']]

mpg  weight
name                                   
chevrolet chevelle malibu  18.0    3504
buick skylark 320          15.0    3693
plymouth satellite         18.0    3436
amc rebel sst              16.0    3433
ford torino                17.0    3449
...                         ...     ...
ford mustang gl            27.0    2790
vw pickup                  44.0    2130
dodge rampage              32.0    2295
ford ranger                28.0    2625
chevrolet s-10             31.0    2720

[398 rows x 2 columns]

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

7. pandas as pd, pd.read_csv("filename"), head()

8. cars.---

9. cars["---"]

10. cars.iloc[---]

11. cars.iloc[row index position, column index position]; "name" column is at index position ... 8

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. cars.---

15. cars.loc["row label"]

16. cars.loc[---, [---, ---]]

17.  cars.loc[ : , [---, ---]]